packages/gatsby/src/utils/html-renderer.js

```diff
@@ -0,0 +1,38 @@
+const Queue = require(`better-queue`)
+const fs = require(`fs-extra`)
+const path = require(`path`)
+
+// copied from https://github.com/markdalgleish/static-site-generator-webpack-plugin/blob/master/index.js#L161
+const generatePathToOutput = outputPath => {
+  let outputFileName = outputPath.replace(/^(\/|\\)/, ``) // Remove leading slashes for webpack-dev-server
+
+  if (!/\.(html?)$/i.test(outputFileName)) {
+    outputFileName = path.join(outputFileName, `index.html`)
+  }
+
+  return path.join(process.cwd(), `public`, outputFileName)
+}
+
+module.exports = (htmlComponentRenderer, pages) =>
+  new Promise(resolve => {
+    const queue = new Queue(
```

Comment:
`p-queue` may be a good option for promise support?

Reply:
I will have a look at it tomorrow

Reply:
p-queue is nice but better-queue is *better* in that it gives us a lot more knobs and buttons to play with e.g. support for merging, filtering, etc. We're also using it a number of other places in core & plugins.

Reply:
better-queue also gathers a lot of stats which would be interesting for analysis https://www.npmjs.com/package/better-queue#queue-statistics

Reply:
Another idea I came across today is to use http://bluebirdjs.com/docs/api/promise.map.html#promise.map . I don't think we actually need any of the knobs apart from concurrency setting (at least right now) and using that would get rid of some of the boilerplate code here and doesn't require to add additional dependency to gatsby

Reply:
I'd rather have the optionality to tweak more things + I want to add an optional build stats report and the build-html queue stats would be really valuable thing to add.

Reply:
Oh also we're already using better-queue in core so this isn't a new dependency.

Reply:
I meant `p-queue` as potential new dependency. But I guess this can just stay as is then. Unless you would like me to add html build stats in this PR?

Reply:
Oh right, sorry.

No the stats can come in a future PR — let's get this in so our first beta will be in great shape. We need to think some first about the design of the stats gathering.